### PR TITLE
externalconn: implement `SHOW EXTERNAL CONNECTIONS`.

### LINKED ISSUE #85905: externalconn: implement `SHOW EXTERNAL CONNECTIONS`
`SHOW EXTERNAL CONNECTIONS` will display information about the external connections that a user has the `USAGE` privilege on. The connection details for each external connection provider must be redacted and so every provider will be responsible for registeringa redaction method that will be invoked before rendering the output for `SHOW EXTERNAL CONNECTIONS`.

Epic [CRDB-15001](https://cockroachlabs.atlassian.net/browse/CRDB-15001)

Jira issue: CRDB-18485

### PR DESCRIPTION
**Pull request by shubhamdhama**
This PR introduces a new query that displays redacted URIs of external connections along with other valuable information.

Epic: CRDB-15001
Fixes: #85905
Release note: New queries `SHOW EXTERNAL CONNECTIONS` and `SHOW EXTERNAL CONNECTION <connection name>` have been added. These queries display redacted connection URIs and other useful information such as the connection type. Access to these queries is restricted to the owner of the connection or users with `USAGE` privilege. Admin or root users will have unrestricted access to all connections.

### CODE REVIEW

File: pkg/ccl/cloudccl/externalconn/testdata/multi-tenant/privileges_show_external_connections
Comment (stevendanna) [inline pkg/ccl/cloudccl/externalconn/testdata/multi-tenant/privileges_show_external_connections:1]: Can we put this test in `pkg/sql/logictest/testdata/logic_test/external_connection_privileges ` instead?

**Review by stevendanna — COMMENTED** [review]
@shubhamdhama Nice work figuring out our testing setup and getting things going.

I think we've had a little bit of confusion. We don't want to add columns to `SHOW CREATE EXTERNAL CONNECTION` because we want that command to be consistent with other `SHOW CREATE` commands.

Rather, we wanted a new command `SHOW EXTERNAL CONNECTION` that has our new output.

File: pkg/sql/show_external_connection.go
Comment (stevendanna) [inline pkg/sql/show_external_connection.go:58]: You can do: `rows := make([]tree.Datums, 0, len(connections))` but I might go further and do `v := p.newContainerValuesNode(showExternalConnectionsColumns, len(connections))` and iterate over connections directly rather than making two arrays.

File: pkg/sql/show_external_connection.go
Comment (stevendanna) [inline pkg/sql/show_external_connection.go:46]: This is fine for now, but one thing to point out is that this will require the entire SHOW output to fit into memory.  That is already a problem with `SHOW CREATE` so is a fine place to start.

File: pkg/sql/show_external_connection.go
Comment (stevendanna) [inline pkg/sql/show_external_connection.go:33]: [nit] We try to comment public functions.

File: pkg/sql/show_create_external_connection.go
Comment (stevendanna) [inline pkg/sql/show_create_external_connection.go:42]: If you wanted you could initialise this with the size of `rows` (`make([]string, 0, len(rows)`) below, closer to where it is used, under the assumption that the common case is that the user has permission.

File: pkg/sql/show_create_external_connection.go
Comment (stevendanna) [inline pkg/sql/show_create_external_connection.go:79]: 🤔 I wonder if we should return an error in the case where ConnectionLabel != nil.

**Review by stevendanna — COMMENTED** [review]
Nice to see this moving along!

File: pkg/sql/show_create_external_connection.go
Comment (shubhamdhama) [reply pkg/sql/show_create_external_connection.go:84]: Agree!

File: pkg/sql/logictest/testdata/logic_test/show_external_connections
Comment (shubhamdhama) [inline pkg/sql/logictest/testdata/logic_test/show_external_connections:3]: Highlighting this comment. Creating external connection wasn't possible here so I placed the tests for URI redaction at the above mentioned location.

**Comment by shubhamdhama** [comment]
(_Moving from description to a comment_) 
Checklist before merging:
- [x] Show redacted connection URIs. Currently, just added two columns.
- [ ] Restrict access to this query based on privileges (subject to discussion).
- [x] Make sure owners have privilege to check their own connections.
- [ ] Update `docs/generated/sql/bnf/BUILD.bazel` to fix failing test.

**Review by stevendanna — COMMENTED** [review]
LGTM, feel free to move from draft. That'll probably pull in a few SQL reviewers who know more about the dark corners of adding new syntax.

As discussed, we think that as a follow-up PR we may want to make this not require buffering all external conns in memory, but for now this follows the same pattern as SHOW CREATE EXTERNAL CONNECTIONS.

**Comment by shubhamdhama** [comment]
One follow-up for this could be to introduce (or reuse an existing) a privilege that would allow a non-admin user to see all external connections instead of current behavior where admin would grant `usage` privilege individually for each connection and user.

**Review by stevendanna — APPROVED** [review]
Overall this looks good to me with the noted follow ups.  Probably good to have a second review either from SQL or CDC since we've touched some of their code here.

Since you've squashed the commits, mind putting the text of the PR description in the commit message? This is nice means that the first thing you find via `git-blame` has the relevant context.

@shubhamdhama Mind searching for / filing issues for those follow-up tasks?

**Comment by shubhamdhama** [comment]
Sure. I'll wait for review from other reviewers before filing those issues. 
As for commit message, I think I have done that (correct me if I'm missing something).

**Review by rafiss — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123545)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shubhamdhama, @stevendanna, and @wenyihu6)

___
*[`pkg/sql/show_external_connection.go` line 32 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123545#-Nxn2BX37GuPqSrQD5RI:-Nxn2BX37GuPqSrQD5RJ:bn6jpwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe4c065c2c26e713f7b130f854c58a7906bf3a18/pkg/sql/show_external_connection.go#L32)):*
> ```Go
> 	extConnNameIdx = iota
> 	extConnURIIdx
> 	extConnTypeIdx
> ```

i don't think these need to be exposed as package-wide constants. instead, these constants can be declared inside of the `ShowExternalConnection` function, which is the only function that references these constants.
___
*[`pkg/sql/show_external_connection.go` line 52 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123545#-Nxn1So6EhBGecF-Rwfi:-Nxn1So6EhBGecF-Rwfj:b-d0dj0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe4c065c2c26e713f7b130f854c58a7906bf3a18/pkg/sql/show_external_connection.go#L52)):*
> ```Go
> 		constructor: func(ctx context.Context, p *planner) (planNode, error) {
> 			params := runParams{ctx: ctx, p: p, extendedEvalCtx: &p.extendedEvalCtx}
> 			connectionNames, err := getConnectionNames(params, n.ConnectionLabel, true)
> ```

super nit: this code might be easier to work with if the privilege check was done here in this function while iterating in the loop below, rather than passing a boolean flag to `getConnectionNames` that turns privilege checking on/off. this is up to you / the DR team though.

another convention that can be useful for improving readability is to add a comment such as:
```
connectionNames, err := getConnectionNames(params, n.ConnectionLabel, true /* checkUsagePrivilege */)
```

**Comment by shubhamdhama** [comment]
*[`pkg/sql/show_external_connection.go` line 32 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123545#-Nxn2BX37GuPqSrQD5RI:-NxpqOLr5DeFcL-dMVrU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe4c065c2c26e713f7b130f854c58a7906bf3a18/pkg/sql/show_external_connection.go#L32)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't think these need to be exposed as package-wide constants. instead, these constants can be declared inside of the `ShowExternalConnection` function, which is the only function that references these constants.
</blockquote></details>

Done.

**Comment by shubhamdhama** [comment]
*[`pkg/sql/show_external_connection.go` line 52 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123545#-Nxn1So6EhBGecF-Rwfi:-NxpqWEQ5wEv8-sNR2zp:bzd0zav) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe4c065c2c26e713f7b130f854c58a7906bf3a18/pkg/sql/show_external_connection.go#L52)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: this code might be easier to work with if the privilege check was done here in this function while iterating in the loop below, rather than passing a boolean flag to `getConnectionNames` that turns privilege checking on/off. this is up to you / the DR team though.

another convention that can be useful for improving readability is to add a comment such as:
```
connectionNames, err := getConnectionNames(params, n.ConnectionLabel, true /* checkUsagePrivilege */)
```
</blockquote></details>

I agree, in fact that's how I ended up splitting `loadExternalConnections` into `getConn..`.  Though moving to below loop would mean that `loadExternalConnections` is fetching details for connections on which we may not have the usage privilege. But that's fine as we can filter `connectionNames` between the calls to these two functions. But I think I'm going to defer this to the follow-up issue where we want to change `QueryBufferedExWithCols` to the iterator based one.

Addressed the second comment.

**Review by stevendanna — APPROVED** [review]

**Comment by shubhamdhama** [comment]
bors r= stevendanna,rafi

**Pull request merged**
